pimcore/models/Tool/CustomReport/Config.php

```diff
@@ -150,7 +150,23 @@ public static function getReportsList()
     public static function getAdapter($configuration, $fullConfig = null)
     {
         $type = $configuration->type ? ucfirst($configuration->type) : 'Sql';
-        $adapter = "\\Pimcore\\Model\\Tool\\CustomReport\\Adapter\\{$type}";
+        if ($type === 'Custom') {
+            if (@\class_exists($configuration->className, true)) { // can need autoloader, can send warning and break json response so we use @
+                $adapter = $configuration->className;
+                $adpaterInstance = new $adapter($configuration, $fullConfig);
+                // we make some check for security reason (because className can be choosen at the GUI level)
+                if (get_parent_class($adpaterInstance) !== 'Pimcore\\Model\\Tool\\CustomReport\\Adapter\\AbstractAdapter') {
```

Comment:
Thanks mrudnicki. You are right we can use instanceof. It allows more felxibility as AbstractAdapter could be any parent of the class. Purpose in my code was to force it to be the direct parent, just to enforce security as giving the user to inject directly the class name to be call from the GUI.
If you think this is not a security problem so instanceof is better, and you can change the code to it yes.